File: python/Binary Tree Preorder Traversal.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def preorderTraversal(self, root):
        stack, answer = [root], []
        while stack:
            top = stack.pop()
            if top:
                answer.append(top.val)
                stack.append(top.right)
                stack.append(top.left)
        return answer
        

File: cpp/Binary Tree Preorder Traversal.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    vector<int> res;
    vector<int> preorderTraversal(TreeNode* root) {
        helper(root);
        return res;
    }
    void helper(TreeNode* root) {
        if (root == NULL) {
            return;
        }
        res.push_back(root->val);
        helper(root->left);
        helper(root->right);
    }
};
